src/librbd/image/RemoveRequest.cc

```diff
@@ -0,0 +1,905 @@
+#include "librbd/operation/TrimRequest.h"
+
+#define dout_subsys ceph_subsys_rbd
+#undef dout_prefix
+#define dout_prefix *_dout << "librbd::image::RemoveRequest: "
+
+namespace librbd {
+namespace image {
+
+using util::create_context_callback;
+using util::create_rados_ack_callback;
+
+template<typename I>
+RemoveRequest<I>::RemoveRequest(IoCtx &ioctx,
+                                const std::string &image_name,
+                                const std::string &image_id,
+                                ProgressContext &prog_ctx, bool force,
+                                ContextWQ *op_work_queue, Context *on_finish) :
+  m_image_name(image_name), m_image_id(image_id), m_prog_ctx(prog_ctx),
+  m_force(force), m_snap_lock("RemoveRequest::m_snap_lock"),
+  m_op_work_queue(op_work_queue), m_on_finish(on_finish) {
+  m_ioctx.dup(ioctx);
+  m_cct = reinterpret_cast<CephContext *>(m_ioctx.cct());
+
+  m_image_ctx = new I((m_image_id.empty() ? m_image_name : std::string()),
+                      m_image_id, nullptr, m_ioctx, false);
+}
+
+template<typename I>
+void RemoveRequest<I>::send() {
+  ldout(m_cct, 20) << this << " " << __func__ << dendl;
+
+  open_image();
+}
+
+template<typename I>
+void RemoveRequest<I>::open_image() {
+  ldout(m_cct, 20) << this << " " << __func__ << dendl;
+
+  using klass = RemoveRequest<I>;
+  Context *ctx = create_context_callback<klass, &klass::handle_open_image>(this);
+
+  m_image_ctx->state->open(ctx);
+}
+
+template<typename I>
+Context *RemoveRequest<I>::handle_open_image(int *result) {
+  ldout(m_cct, 20) << __func__ << ": r=" << *result << dendl;
+
+  if (*result < 0) {
+    lderr(m_cct) << "error opening image: " << cpp_strerror(*result) << dendl;
+    send_close_image(*result);
+    return nullptr;
+  }
+
+  m_image_id = m_image_ctx->id;
+  m_image_name = m_image_ctx->name;
+  m_header_oid = m_image_ctx->header_oid;
+  m_old_format = m_image_ctx->old_format;
+  m_unknown_format = false;
+
+  check_exclusive_lock();
+  return nullptr;
+}
+
+template<typename I>
+void RemoveRequest<I>::check_exclusive_lock() {
+  ldout(m_cct, 20) << this << " " << __func__ << dendl;
+
+  m_image_ctx->owner_lock.get_read();
+  if (m_image_ctx->exclusive_lock == nullptr) {
+    validate_image_removal();
+  } else {
+    acquire_exclusive_lock();
+  }
+}
+
+template<typename I>
+void RemoveRequest<I>::acquire_exclusive_lock() {
+  ldout(m_cct, 20) << this << " " << __func__ << dendl;
+
+  using klass = RemoveRequest<I>;
+  if (m_force) {
+    Context *ctx =
+      create_context_callback<klass, &klass::handle_exclusive_lock_force>(this);
+    // releasing read lock to avoid a deadlock when upgrading to
+    // write lock in the shut_down process
+    m_image_ctx->owner_lock.put_read();
+    m_image_ctx->exclusive_lock->shut_down(ctx);
+  } else {
+    Context *ctx = create_context_callback<klass, &klass::handle_exclusive_lock>(this);
+    m_image_ctx->operations->prepare_image_update(ctx);
```

Comment:
You can just directly request the lock -- no need to use this sync helper function